File: data-structures/binary-heap/binary-heap.go
package main

import (
	"errors"
	"fmt"
)

const MaxUint = ^uint(0)         // uint of 0 is 64 0s when you complement it you get 64 1s hence the max uint value
const MaxInt = int(MaxUint >> 1) // shift one bit to right 64 64 bits now become 32 and then convert it to int and Voila! you have max int value
type MaxBinaryHeap struct {
	values []int
}

func (h *MaxBinaryHeap) insert(value int) {
	h.values = append(h.values, value)
	h.bubbleUp()
}

func (h *MaxBinaryHeap) bubbleUp() {
	idx := len(h.values) - 1
	element := h.values[idx]

	for idx > 0 {
		parentIdx := ((idx - 1) / 2)
		parent := h.values[parentIdx]
		if element <= parent {
			break
		}

		h.values[parentIdx] = element
		h.values[idx] = parent
		idx = parentIdx

	}
}

func (h *MaxBinaryHeap) extractMax() (value int, err error) {
	if len(h.values) > 0 {
		root := h.values[0]
		end := h.values[len(h.values)-1]
		h.values = h.values[:len(h.values)-1]

		if len(h.values) > 1 {
			h.values[0] = end
			h.bubbleDown()
		}
		return root, nil
	}
	return 0, errors.New("no element to extract")

}

func (h *MaxBinaryHeap) bubbleDown() {
	parent := h.values[0]
	parentIdx := 0
	for true {
		leftIdx := (2 * parentIdx) + 1
		rightIdx := (2 * parentIdx) + 2
		var leftChild, rightChild, swapIdx interface{}
		if leftIdx < len(h.values) {
			leftChild = h.values[leftIdx]
			if leftChild.(int) > parent {
				swapIdx = leftIdx
			}
		}
		if rightIdx < len(h.values) {
			rightChild = h.values[rightIdx]
			if (swapIdx == nil && rightChild.(int) > parent) || (swapIdx != nil && rightChild.(int) > leftChild.(int)) {
				swapIdx = rightIdx
			}
		}

		if swapIdx == nil {
			break
		}

		h.values[parentIdx] = h.values[swapIdx.(int)]
		h.values[swapIdx.(int)] = parent
		parentIdx = swapIdx.(int)
	}
}

func (h *MaxBinaryHeap) deleteKey(index int) (value int, err error) {
	if index > len(h.values) {
		return 0, errors.New("out of range")
	}
	elementToBeDeleted := h.values[index]
	h.values[index] = MaxInt
	idx := index
	element := h.values[idx]

	for idx != 0 {
		parentIdx := ((idx - 1) / 2)

		parent := h.values[parentIdx]

		h.values[parentIdx] = element
		h.values[idx] = parent
		idx = parentIdx
	}
	h.extractMax()
	return elementToBeDeleted, nil
}

func main() {
	heap := MaxBinaryHeap{}
	heap.insert(87)
	heap.insert(72)
	heap.insert(70)
	heap.insert(55)
	heap.insert(58)
	heap.insert(47)
	heap.insert(42)
	heap.insert(35)
	heap.insert(34)
	heap.insert(26)
	heap.insert(25)
	heap.insert(38)
	heap.insert(39)
	heap.extractMax()
	fmt.Println(heap.values)
}

File: data-structures/binary-heap/binary-heap.js
class MaxBinaryHeap {
    constructor() {
        this.values = []
    }

    insert(value) {
        this.values.push(value)
        this.bubbleUp()

    }

    bubbleUp() {
        var idx = this.values.length - 1;
        var element = this.values[idx];

        while (idx > 0) {
            var parentIdx = Math.floor((idx - 1) / 2);
            var parent = this.values[parentIdx];
            if (element <= parent)
                break;

            this.values[parentIdx] = element;
            this.values[idx] = parent;
            idx = parentIdx;

        }

    }

    extractMax() {
        var root = this.values[0];
        var end = this.values.pop();
        if (this.values.length > 0) {
            this.values[0] = end;
            this.bubbleDown()
        }

        return root;

    }

    bubbleDown() {
        var parent = this.values[0]
        var parentIdx = 0;
        while (true) {
            var leftIdx = (2 * parentIdx) + 1;
            var rightIdx = (2 * parentIdx) + 2;
            var leftChild, rightChild, swapIdx = null;

            if (leftIdx < this.values.length) {
                leftChild = this.values[leftIdx];
                if (leftChild > parent) {
                    swapIdx = leftIdx;
                }
            }
            if (rightIdx < this.values.length) {
                rightChild = this.values[rightIdx];
                if ((swapIdx === null && rightChild > parent) || (swapIdx !== null && rightChild > leftChild)) {
                    swapIdx = rightIdx
                }
            }

            if (swapIdx === null) {
                break
            }

            this.values[parentIdx] = this.values[swapIdx];
            this.values[swapIdx] = parent;
            parentIdx = swapIdx;
        }
    }

    deleteKey(index) {
        if (index > this.values.length) {
            return null;
        }
        var elementToBeDeleted = this.values[index];
        this.values[index] = Infinity;
        var idx = index;
        var element = this.values[idx];

        while (idx !== 0) {
            var parentIdx = Math.floor((idx - 1) / 2);
            console.log(parentIdx);
            var parent = this.values[parentIdx];

            this.values[parentIdx] = element;
            this.values[idx] = parent;
            idx = parentIdx;
        }
        this.extractMax();
        return elementToBeDeleted;

    }

}

var heap = new MaxBinaryHeap();
heap.insert(87)
heap.insert(72)
heap.insert(70)
heap.insert(55)
heap.insert(58)
heap.insert(47)
heap.insert(42)
heap.insert(35)
heap.insert(34)
heap.insert(26)
heap.insert(25)
heap.insert(38)
heap.insert(39)
